text/1008-tls-and-alpn.rst

```diff
@@ -0,0 +1,458 @@
+certificates using a local CA, and the CLI tells the server to trust the
+generated certificates.
+
+The certificate should contain the authorized database role in CN or an
+X.509 extension, and that role must match the requested login user
+during authentication. As the server may support several different Auth
+methods at the same time with a customizable priority, a client
+certificate is not mandatory in TLS. But if provided and if the server
+is configured with ``Certificate`` Auth, then the client certificate
+will be used as one authentication attempt.
+
+On the client side, user may use the CLI to generate a client
+certificate (and its corresponding private) for a particular database
+role in a certain EdgeDB instance, and use the two files to establish a
+connection to that EdgeDB server. The private key passphrase - if set -
+must be securely provided through either environment variables, or API
+parameters (following Python ``SSLContext.load_cert_chain()`` style).
+We may be able to place the client certificate in the
+``credentials.json`` file so that the user don't have to bother dealing
+with the certificates any more. And we could likely skip the passphrase
+for development client certificates.
+
+
+Backwards Compatibility
+=======================
+
+While TLS will be enforced by default, compatible mode is still
+available for the server before EdgeDB 1.0, but it is only for the
+EdgeDB developers and should not be enabled by the users.
+
++------------+----------------+----------------+---------------------------+
+|            | Old Server     | New Server     | New Server in Compat Mode |
++============+================+================+===========================+
+| Old Client | Accessible     | Friendly Error | Accessible                |
++------------+----------------+----------------+---------------------------+
+| New Client | Friendly Error | Accessible     | Friendly Error            |
++------------+----------------+----------------+---------------------------+
+
+The EdgeDB development server (``edb server``) will provide a hidden
+option ``--no-tls`` to run the server in non-TLS compatible mode (just
+like the old server) for development and testing only. This option is
+not available in the EdgeDB CLI (``edgedb server``).
+
+At the same time, the new server will return a user-friendly error in
+plain text if the SSL handshake fails, in binary protocol or HTTP
+depending on again the magical first-byte. Similarly, if the new client
+could not establish a TLS connection, it should raise a proper error
+with the reason.
+
+
+Security Implications
+=====================
+
+Enforcing TLS is supposed to be a full level-up in terms of security. It
+provides basic eavesdropping protection, and if configured properly the
+MITM protection too. This needs to be carefully documented as this RFC
+does not involve setting up a proper CA-based trust chain for MITM
+attacks.
```

Comment:
You don't need CA-based trust to prevent MITM. Veryfying server certificate is enough. So generating `credentials.json` by web UI of cloud (Aether) or providing certificate fingerprint alongside of user and password from DB Ops to Users within the company is enough for that. Client certificates are even better.